### DISCUSSION TITLE
Surface Laptop 4 AMD wiki amd_iommu

### DISCUSSION THREAD

**Discussion by darbyhaller** [linux-surface/linux-surface#480] [Ideas]
https://github.com/linux-surface/linux-surface/wiki/Surface-Laptop-4 states 

> Setting amd_iommu=force_isolation appears to result in a functional system, a crash has not yet been observed.
>Alternatively, nomodeset will cause the amdgpu driver to be prevented from loading. Disabling the IOMMU security feature with amd_iommu=off iommu=off also appears to work. Setting pci=noats the system appears to function for a while, but graphics locks up later, maybe PM related, tends to occur after screen turning off.

However, I experience crashes with amd_iommu=force_isolation when using graphically intensive applications like Zoom or google maps. I use amd_iommu=off iommu=off instead, which work well. But I'm not knowledgeable about what exactly these amd_iommu options are doing, so I don't know if I should edit the wiki page to reflect this myself or not.

**Comment by qzed** [comment]
v5.12.12 should improve that situation a bit. If crashes still occur with that, maybe change the two sentences around indicating that `amd_iommu=off iommu=off` should be the preferred fix for now. I don't think iommu brings that much more security (though I might be wrong, I only really know that from the context of hardware passthrough in VMs).

**Comment by nakato** [comment]
I've updated the wiki to prefer disabling the iommu for the SL4A.

**Reply by ghost** [reply]
amd_iommu=off is enough, it overrides iommu :)

**Comment by dperish** [comment]
Pardon the newbie question, but I'm not sure how to apply these settings.  Currently running on SL4/AMD with Mint 20.02/Cinnamon 64bit, in Window 11 dual boot setup.

**Reply by dperish** [reply]
This seems to have done the trick for me:

`sudo nano /etc/default/grub` 

Added the parameters to the following line:

`GRUB_CMDLINE_LINUX="amd_iommu=off iommu=off"`

`sudo update-grub`

Then rebooted cleanly without the display corruption =)

**Reply by DrBrad** [reply]
Doing what you have shown causes the computer to be inoperable after a hibernate and suspend within Ubuntu.
Although it allows you to boot without corruption you wont be able to hibernate and suspend.

**Reply by nakato** [reply]
`amd_iommu=off iommu=off` is what I'm booted with, sleep (S2Idle/s0ix) has been working for me, I've been using the 5.15.11-arch2-1-surface kernel from here for the last week, uptime 6 days 20 hours.

I do not have hibernation support in my setup however, and I cannot test hibernation without re-partitioning and re-formatting my disk.

**Reply by dperish** [reply]
Having the same issue as @DrBrad in Mint 20.03.  I didn't see an option for hibernation in the BIOS, and haven't done anything in Mint to explicitly enable it.  

After looking at the Mint forums (https://forums.linuxmint.com/viewtopic.php?t=357914), it looks like I'd need 2x the swap space allocated (current only have 8gb vs 16gb of physical ram), and would also have to disable secure boot.

@nakato does your arch distribution have secure boot disabled, and could that be the difference here?

**Reply by nakato** [reply]
>  didn't see an option for hibernation in the BIOS

The Microsoft firmware has nearly no options available to us end-users, it's configured however it needs to be configured to be optimal for Windows.  Hibernation is on in the firmware.

> haven't done anything in Mint to explicitly enable it.

It is almost certainly enabled by default to some extent.  `cat /sys/power/disk` will probably show *platform* selected, which is some cooperation with the firmware for hibernation.
There are some potential toggles you can use to disable hibernate here:  https://www.freedesktop.org/software/systemd/man/systemd-sleep.conf.html

It would be good to know if resume-from-sleep is failing on it's own or if it's part of hibernate.  To isolate this you'd need disable your graphical interface, and try suspending with `echo 'mem' >/sys/power/state`, and bring it in and out of s0ix a few times.

Keep in mind the laptop will only wake from a few inputs though.  A lid event or a tap of the power button should work; no other key on the keyboard will wake it.
Also, if laptop-mode-tools, powertop, or anything else has run, you'll need to disable them and reboot first to isolate the changes they make as a potential cause of the issues.

> does your arch distribution have secure boot disabled, and could that be the difference here?

It's disabled, but I can't hibernate, I don't have swap and I have all hibernation configuration disabled in systemd-sleep.conf, this is probably the difference.

**Reply by DrBrad** [reply]
I have tried doing 20GB of swap on my laptop, however not double the memory, I'll test it and see if it works.

**Reply by dperish** [reply]
@nakato thanks for all the info, it was very helpful to me to understand the configuration and options in the arena.  

Unfortunately, after disabling secure boot & swap, and adding the following to `/etc/systemd/sleep.conf`, it's still exhibiting the same behavior where I had to long press the power button and then tap it a few more times to initiate a cold boot:

```
[Sleep]
AllowSuspend=yes
AllowHibernation=no
AllowSuspendThenHibernate=no
AllowHybridSleep=no
SuspendMode=suspend
SuspendState=standby freeze
```

FYI, here's the contents from `/sys/power/disk`:
```
$ cat /sys/power/disk
[platform] shutdown reboot suspend test_resume 
```

**Reply by DrBrad** [reply]
https://github.com/linux-surface/linux-surface/issues/624

**Comment by Kaffarov** [comment]
The camera for surface pro 4 doesn't work on kali linux. Who to fix this?
